File: backend/app/api/conversation_api.py
"""Agent 对话历史 HTTP API（会话 + 消息；与 Agent 执行内核解耦，由调用方按需写入/读取）。

多轮与 Agent 编排的衔接见 ``AgentInvokeRequest`` / ``AgentInvokeData``：须在调用 ``/api/agent/invoke`` 或流式接口时
回传 ``conversation_session_id``，否则仅传 ``agent_id`` 时每次都会新建会话。
"""

from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, Query

from app.core.constants.conversation import (
    CONVERSATION_DETAIL_MESSAGES_DEFAULT_PAGE_SIZE,
    CONVERSATION_DETAIL_MESSAGES_MAX_PAGE_SIZE,
    CONVERSATION_LIST_DEFAULT_PAGE_SIZE,
    CONVERSATION_LIST_MAX_PAGE_SIZE,
)
from app.core.deps import DBManagerDep, OptionalRedisDep
from app.domain.conversation import ConversationDomainError
from app.schemas.conversation import (
    ConversationDetailData,
    ConversationMessageAppendBody,
    ConversationMessageExecutionOut,
    ConversationMessageOut,
    ConversationSessionCreateBody,
    ConversationSessionListData,
    ConversationSessionOut,
    ConversationSessionPatchBody,
    RollingSummaryEnqueueOut,
)
from app.schemas.response import ApiResponse
from app.services.conversation_svc import ConversationService


def get_conversation_service(
    db_manager: DBManagerDep,
    redis: OptionalRedisDep,
) -> ConversationService:
    return ConversationService(db_manager, redis=redis)


ConversationServiceDep = Annotated[ConversationService, Depends(get_conversation_service)]


router = APIRouter()


@router.post(
    "/sessions",
    response_model=ApiResponse[ConversationSessionOut],
    summary="创建对话会话",
    operation_id="post_conversation_session",
)
async def create_session(
    body: ConversationSessionCreateBody,
    svc: ConversationServiceDep,
) -> ApiResponse[ConversationSessionOut]:
    try:
        data = await svc.create_session(body)
    except ConversationDomainError as e:
        raise HTTPException(status_code=e.status_code, detail=str(e)) from e
    return ApiResponse(message="ok", data=data)


@router.get(
    "/sessions",
    response_model=ApiResponse[ConversationSessionListData],
    summary="分页列出对话会话（不含已软删除）",
    operation_id="get_conversation_sessions",
)
async def list_sessions(
    svc: ConversationServiceDep,
    page: Annotated[int, Query(ge=1)] = 1,
    page_size: Annotated[
        int,
        Query(ge=1, le=CONVERSATION_LIST_MAX_PAGE_SIZE),
    ] = CONVERSATION_LIST_DEFAULT_PAGE_SIZE,
    agent_id: Annotated[int | None, Query(description="按 agent_entity.id 筛选")] = None,
) -> ApiResponse[ConversationSessionListData]:
    data = await svc.list_sessions(
        page=page,
        page_size=page_size,
        agent_id=agent_id,
    )
    return ApiResponse(message="ok", data=data)


@router.post(
    "/sessions/{session_id}/rolling-summary",
    response_model=ApiResponse[RollingSummaryEnqueueOut],
    summary="手动触发滚动摘要（异步入队）",
    operation_id="post_conversation_session_rolling_summary",
)
async def enqueue_rolling_summary(
    session_id: str,
    svc: ConversationServiceDep,
) -> ApiResponse[RollingSummaryEnqueueOut]:
    """将摘要任务写入进程内队列；实际 LLM 在后台执行。"""
    try:
        outcome = await svc.request_manual_rolling_summary(session_id)
    except LookupError as e:
        raise HTTPException(status_code=404, detail=str(e)) from e
    if outcome == "duplicate":
        return ApiResponse(
            message="已有摘要任务在排队或执行中",
            data=RollingSummaryEnqueueOut(queued=False),
        )
    if outcome == "queue_unavailable":
        raise HTTPException(
            status_code=503,
            detail="摘要队列未就绪（服务未启动 RollingSummaryWorker）",
        ) from None
    return ApiResponse(message="ok", data=RollingSummaryEnqueueOut(queued=True))


@router.get(
    "/sessions/{session_id}",
    response_model=ApiResponse[ConversationDetailData],
    summary="按会话 ID 获取会话元数据与消息分页（第 1 页为最近一批，页内按时间正序）",
    operation_id="get_conversation_session_detail",
)
async def get_session_detail(
    session_id: str,
    svc: ConversationServiceDep,
    page: Annotated[int, Query(ge=1)] = 1,
    page_size: Annotated[
        int,
        Query(ge=1, le=CONVERSATION_DETAIL_MESSAGES_MAX_PAGE_SIZE),
    ] = CONVERSATION_DETAIL_MESSAGES_DEFAULT_PAGE_SIZE,
    messages_agent_id: Annotated[
        int | None,
        Query(
            description="仅返回并统计 agent_id 等于该值的会话消息；工作台历史可传当前 workbench 的 agent_entity.id",
        ),
    ] = None,
    include_message_metadata: Annotated[
        bool,
        Query(
            description=(
                "为 false 时不加载各消息 metadata JSON、响应中 metadata 恒为 null；"
                "执行过程等用 GET .../messages/{message_id}/execution"
            ),
        ),
    ] = True,
) -> ApiResponse[ConversationDetailData]:
    data = await svc.get_session_detail(
        session_id,
        page=page,
        page_size=page_size,
        messages_agent_id=messages_agent_id,
        include_message_metadata=include_message_metadata,
    )
    if data is None:
        raise HTTPException(status_code=404, detail="会话不存在或已删除") from None
    return ApiResponse(message="ok", data=data)


@router.get(
    "/sessions/{session_id}/messages/{message_id}/execution",
    response_model=ApiResponse[ConversationMessageExecutionOut],
    summary="按消息 ID 查询执行过程快照",
    operation_id="get_conversation_message_execution",
)
async def get_message_execution(
    session_id: str,
    message_id: int,
    svc: ConversationServiceDep,
) -> ApiResponse[ConversationMessageExecutionOut]:
    """读取该条 **助手** 消息 ``metadata`` 中的 ``process_trace`` / ``tool_history`` / ``thinking_text``。"""
    try:
        data = await svc.get_message_execution(session_id, message_id)
    except ConversationDomainError as e:
        raise HTTPException(status_code=e.status_code, detail=str(e)) from e
    return ApiResponse(message="ok", data=data)


@router.post(
    "/sessions/{session_id}/messages",
    response_model=ApiResponse[ConversationMessageOut],
    summary="追加一条消息（活跃会话；关闭后拒绝写入）",
    operation_id="post_conversation_message",
)
async def append_message(
    session_id: str,
    body: ConversationMessageAppendBody,
    svc: ConversationServiceDep,
) -> ApiResponse[ConversationMessageOut]:
    data = await svc.append_message(session_id, body)
    if data is None:
        raise HTTPException(
            status_code=400,
            detail="会话不存在、已删除、已关闭或无法写入",
        ) from None
    return ApiResponse(message="ok", data=data)


@router.patch(
    "/sessions/{session_id}",
    response_model=ApiResponse[ConversationSessionOut],
    summary="更新会话（标题、关闭）",
    operation_id="patch_conversation_session",
)
async def patch_session(
    session_id: str,
    body: ConversationSessionPatchBody,
    svc: ConversationServiceDep,
) -> ApiResponse[ConversationSessionOut]:
    data = await svc.patch_session(session_id, body)
    if data is None:
        raise HTTPException(status_code=404, detail="会话不存在、已删除或状态非法") from None
    return ApiResponse(message="ok", data=data)


@router.delete(
    "/sessions/{session_id}",
    response_model=ApiResponse[dict[str, bool]],
    summary="软删除会话（列表与默认查询不可见）",
    operation_id="delete_conversation_session",
)
async def delete_session(
    session_id: str,
    svc: ConversationServiceDep,
) -> ApiResponse[dict[str, bool]]:
    ok = await svc.soft_delete_session(session_id)
    if not ok:
        raise HTTPException(status_code=404, detail="会话不存在或已删除") from None
    return ApiResponse(message="ok", data={"deleted": True})

File: frontend/src/api/conversationApi.ts
/**
 * Agent 对话历史 API（GET /api/agent-conversations/...，信封 message + data）
 */

import { parseApiEnvelope } from './parseEnvelope'

export type PageMeta = {
  page: number
  page_size: number
  total: number
}

export type ConversationMessageRole = 'user' | 'assistant' | 'system'

export type ConversationSessionOut = {
  session_id: string
  agent_id: number
  title: string
  status: number
  created_at: string
  updated_at: string
  summary?: string | null
  summary_version?: number
  summary_job_status?: number
}

export type ConversationMessageOut = {
  id: number
  session_id: string
  /** 本条消息归属/产生的 Agent；与同 session 下其它 agent 消息区分（像素办公室等多角色同会话） */
  agent_id?: number
  role: ConversationMessageRole
  content: string
  content_type: string
  created_at: string
  metadata: Record<string, unknown> | null
  /** 会话内轮次；0 为占位（system / 历史回填） */
  turn_index: number
  /** assistant 指向本轮 user 消息 id */
  reply_message_id?: number | null
  /** user：正文估算；assistant：网关 usage 总 token */
  tokens?: number | null
}

export type ConversationSessionListData = {
  items: ConversationSessionOut[]
  meta: PageMeta
}

export type ConversationDetailData = {
  session: ConversationSessionOut
  messages: ConversationMessageOut[]
  /** 消息分页：第 1 页为最近一批，页内时间正序 */
  messages_meta: PageMeta
}

/** GET .../messages/{id}/execution — 与后端 ConversationMessageExecutionOut 一致 */
export type ConversationMessageExecutionOut = {
  message_id: number
  session_id: string
  agent_id: number
  role: ConversationMessageRole
  process_trace?: unknown[] | null
  tool_history?: Record<string, unknown>[] | null
  thinking_text?: string | null
}

const jsonHeaders = { 'Content-Type': 'application/json', Accept: 'application/json' }

export type ListConversationSessionsParams = {
  page?: number
  page_size?: number
  agent_id?: number
}

export type ConversationSessionCreateBody = {
  agent_id: number
  title?: string | null
}

/** POST /api/agent-conversations/sessions — 创建空会话（绑定 agent_entity） */
export async function createConversationSession(
  body: ConversationSessionCreateBody,
): Promise<ConversationSessionOut> {
  const res = await fetch('/api/agent-conversations/sessions', {
    method: 'POST',
    headers: jsonHeaders,
    body: JSON.stringify(body),
  })
  return parseApiEnvelope<ConversationSessionOut>(res)
}

export async function listConversationSessions(
  params?: ListConversationSessionsParams,
): Promise<ConversationSessionListData> {
  const q = new URLSearchParams()
  if (params?.page != null) q.set('page', String(params.page))
  if (params?.page_size != null) q.set('page_size', String(params.page_size))
  if (params?.agent_id != null) q.set('agent_id', String(params.agent_id))
  const qs = q.toString()
  const url = `/api/agent-conversations/sessions${qs ? `?${qs}` : ''}`
  const res = await fetch(url, { headers: jsonHeaders })
  return parseApiEnvelope<ConversationSessionListData>(res)
}

export type GetConversationSessionDetailParams = {
  page?: number
  page_size?: number
  /** 仅返回该 agent_id 的会话消息（工作台共享会话下排除子 Agent 行） */
  messages_agent_id?: number
  /**
   * 是否随列表返回每条消息的 metadata（含 process_trace / tool_history 等大字段）。
   * 默认 false：不拉 JSON 列，执行过程用 `getConversationMessageExecution`。
   */
  include_message_metadata?: boolean
}

export async function getConversationSessionDetail(
  sessionId: string,
  params?: GetConversationSessionDetailParams,
): Promise<ConversationDetailData> {
  const q = new URLSearchParams()
  if (params?.page != null) q.set('page', String(params.page))
  if (params?.page_size != null) q.set('page_size', String(params.page_size))
  if (params?.messages_agent_id != null) q.set('messages_agent_id', String(params.messages_agent_id))
  const includeMeta = params?.include_message_metadata === true
  q.set('include_message_metadata', includeMeta ? 'true' : 'false')
  const qs = q.toString()
  const res = await fetch(
    `/api/agent-conversations/sessions/${encodeURIComponent(sessionId)}${qs ? `?${qs}` : ''}`,
    {
      headers: jsonHeaders,
    },
  )
  return parseApiEnvelope<ConversationDetailData>(res)
}

/** GET /api/agent-conversations/sessions/{sessionId}/messages/{messageId}/execution */
export async function getConversationMessageExecution(
  sessionId: string,
  messageId: number,
): Promise<ConversationMessageExecutionOut> {
  const res = await fetch(
    `/api/agent-conversations/sessions/${encodeURIComponent(sessionId)}/messages/${messageId}/execution`,
    { headers: jsonHeaders },
  )
  return parseApiEnvelope<ConversationMessageExecutionOut>(res)
}

export type RollingSummaryEnqueueOut = {
  /** 是否已入队后台摘要任务 */
  queued: boolean
}

/** POST /api/agent-conversations/sessions/{id}/rolling-summary — 手动触发滚动摘要（异步） */
export async function postSessionRollingSummary(
  sessionId: string,
): Promise<RollingSummaryEnqueueOut> {
  const res = await fetch(
    `/api/agent-conversations/sessions/${encodeURIComponent(sessionId)}/rolling-summary`,
    { method: 'POST', headers: jsonHeaders },
  )
  return parseApiEnvelope<RollingSummaryEnqueueOut>(res)
}
